File: backend/events_manager.py
from aws_cdk import DockerImage, Duration, RemovalPolicy
from aws_cdk import aws_appsync_alpha as appsync
from aws_cdk import aws_cognito as cognito
from aws_cdk import aws_dynamodb as dynamodb
from aws_cdk import aws_iam as iam
from aws_cdk import aws_lambda as awslambda
from aws_cdk import aws_lambda_python_alpha as lambda_python
from constructs import Construct


class EventsManager(Construct):
    def __init__(
        self,
        scope: Construct,
        id: str,
        api: appsync.IGraphqlApi,
        none_data_source: appsync.IGraphqlApi,
        user_pool: cognito.IUserPool,
        powertools_layer: lambda_python.PythonLayerVersion,
        powertools_log_level: str,
        lambda_architecture: awslambda.Architecture,
        lambda_runtime: awslambda.Runtime,
        lambda_bundling_image: DockerImage,
        roles_to_grant_invoke_access: list[iam.IRole],
        **kwargs,
    ):
        super().__init__(scope, id, **kwargs)

        events_table = dynamodb.Table(
            self,
            "EventsTable",
            partition_key=dynamodb.Attribute(
                name="eventId", type=dynamodb.AttributeType.STRING
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            encryption=dynamodb.TableEncryption.AWS_MANAGED,
            removal_policy=RemovalPolicy.DESTROY,
        )

        events_handler = lambda_python.PythonFunction(
            self,
            "eventsFunction",
            entry="backend/lambdas/events_function/",
            description="Events Resolver",
            index="index.py",
            handler="lambda_handler",
            timeout=Duration.minutes(1),
            runtime=lambda_runtime,
            tracing=awslambda.Tracing.ACTIVE,
            memory_size=128,
            architecture=lambda_architecture,
            bundling=lambda_python.BundlingOptions(image=lambda_bundling_image),
            layers=[powertools_layer],
            environment={
                "DDB_TABLE": events_table.table_name,
                "user_pool_id": user_pool.user_pool_id,
                "POWERTOOLS_SERVICE_NAME": "events_resolver",
                "LOG_LEVEL": powertools_log_level,
            },
        )

        events_table.grant_read_write_data(events_handler)

        # Define the data source for the API
        events_data_source = api.add_lambda_data_source(
            "EventsDataSource", events_handler
        )

        # Define API Schema

        events_object_Type = appsync.ObjectType(
            "Event",
            definition={
                "eventId": appsync.GraphqlType.id(),
                "createdAt": appsync.GraphqlType.aws_date_time(),
                "eventName": appsync.GraphqlType.string(),
                "eventDate": appsync.GraphqlType.aws_date(),
                "fleetId": appsync.GraphqlType.id(),
                "countryCode": appsync.GraphqlType.string(),
                "raceRankingMethod": appsync.GraphqlType.string(),
                "raceTimeInMin": appsync.GraphqlType.int(),
                "raceNumberOfResets": appsync.GraphqlType.int(),
                "raceLapsToFinish": appsync.GraphqlType.int(),
                "raceTrackType": appsync.GraphqlType.string(),
            },
        )

        api.add_type(events_object_Type)

        # Event methods
        api.add_query(
            "getAllEvents",
            appsync.ResolvableField(
                return_type=events_object_Type.attribute(is_list=True),
                data_source=events_data_source,
            ),
        )
        api.add_mutation(
            "addEvent",
            appsync.ResolvableField(
                args={
                    "eventName": appsync.GraphqlType.string(is_required=True),
                    "eventDate": appsync.GraphqlType.aws_date(),
                    "fleetId": appsync.GraphqlType.id(),
                    "countryCode": appsync.GraphqlType.string(),
                    "raceRankingMethod": appsync.GraphqlType.string(is_required=True),
                    "raceTimeInMin": appsync.GraphqlType.int(is_required=True),
                    "raceNumberOfResets": appsync.GraphqlType.int(is_required=True),
                    "raceLapsToFinish": appsync.GraphqlType.int(is_required=True),
                    "raceTrackType": appsync.GraphqlType.string(is_required=True),
                },
                return_type=events_object_Type.attribute(),
                data_source=events_data_source,
            ),
        )
        api.add_subscription(
            "onAddedEvent",
            appsync.ResolvableField(
                return_type=events_object_Type.attribute(),
                data_source=none_data_source,
                request_mapping_template=appsync.MappingTemplate.from_string(
                    """{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }"""
                ),
                response_mapping_template=appsync.MappingTemplate.from_string(
                    """$util.toJson($context.result)"""
                ),
                directives=[appsync.Directive.subscribe("addEvent")],
            ),
        )

        api.add_mutation(
            "deleteEvents",
            appsync.ResolvableField(
                args={"eventIds": appsync.GraphqlType.string(is_required_list=True)},
                return_type=events_object_Type.attribute(is_list=True),
                data_source=events_data_source,
            ),
        )
        api.add_subscription(
            "onDeletedEvents",
            appsync.ResolvableField(
                return_type=events_object_Type.attribute(is_list=True),
                data_source=none_data_source,
                request_mapping_template=appsync.MappingTemplate.from_string(
                    """{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }"""
                ),
                response_mapping_template=appsync.MappingTemplate.from_string(
                    """$util.toJson($context.result)"""
                ),
                directives=[appsync.Directive.subscribe("deleteEvents")],
            ),
        )

        api.add_mutation(
            "updateEvent",
            appsync.ResolvableField(
                args={
                    "eventId": appsync.GraphqlType.string(is_required=True),
                    "eventName": appsync.GraphqlType.string(),
                    "eventDate": appsync.GraphqlType.aws_date(),
                    "fleetId": appsync.GraphqlType.id(),
                    "countryCode": appsync.GraphqlType.string(),
                    "raceTrackType": appsync.GraphqlType.string(),
                    "raceRankingMethod": appsync.GraphqlType.string(),
                    "raceTimeInMin": appsync.GraphqlType.int(),
                    "raceNumberOfResets": appsync.GraphqlType.int(),
                    "raceLapsToFinish": appsync.GraphqlType.int(),
                },
                return_type=events_object_Type.attribute(),
                data_source=events_data_source,
            ),
        )
        api.add_subscription(
            "onUpdatedEvent",
            appsync.ResolvableField(
                return_type=events_object_Type.attribute(),
                data_source=none_data_source,
                request_mapping_template=appsync.MappingTemplate.from_string(
                    """{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }"""
                ),
                response_mapping_template=appsync.MappingTemplate.from_string(
                    """$util.toJson($context.result)"""
                ),
                directives=[appsync.Directive.subscribe("updateEvent")],
            ),
        )

        # Grant access so API methods can be invoked
        for role in roles_to_grant_invoke_access:
            role.add_to_policy(
                iam.PolicyStatement(
                    effect=iam.Effect.ALLOW,
                    actions=["appsync:GraphQL"],
                    resources=[
                        f"{api.arn}/types/Query/fields/getAllEvents",
                        f"{api.arn}/types/Mutation/fields/addEvent",
                        f"{api.arn}/types/Subscription/fields/addedEvent",
                        f"{api.arn}/types/Mutation/fields/deleteEvent",
                        f"{api.arn}/types/Subscription/fields/deletedEvent",
                        f"{api.arn}/types/Mutation/fields/updateEvent",
                        f"{api.arn}/types/Subscription/fields/addedEvent",
                        f"{api.arn}/types/Subscription/fields/deletedEvent",
                        f"{api.arn}/types/Subscription/fields/updatedEvent",
                        # f'{api.arn}/types/Mutation/fields/addTrack',
                        # f'{api.arn}/types/Mutation/fields/deleteTrack',
                        # f'{api.arn}/types/Mutation/fields/updateTrack',
                    ],
                )
            )

File: lib/constructs/events-manager.ts
import { DockerImage, Duration, RemovalPolicy, Stack } from 'aws-cdk-lib';
import * as appsync from 'aws-cdk-lib/aws-appsync';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { EventBus } from 'aws-cdk-lib/aws-events';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { StartingPosition } from 'aws-cdk-lib/aws-lambda';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import {
  CodeFirstSchema,
  Directive,
  EnumType,
  GraphqlType,
  InputType,
  ObjectType,
  ResolvableField,
} from 'awscdk-appsync-utils';
import { NagSuppressions } from 'cdk-nag';
import { StandardLambdaPythonFunction } from './standard-lambda-python-function';

import { Construct } from 'constructs';

export interface EventsManagerProps {
  appsyncApi: {
    schema: CodeFirstSchema;
    api: appsync.GraphqlApi;
    noneDataSource: appsync.NoneDataSource;
  };
  lambdaConfig: {
    runtime: lambda.Runtime;
    architecture: lambda.Architecture;
    bundlingImage: DockerImage;
    layersConfig: {
      powerToolsLogLevel: string;
      helperFunctionsLayer: lambda.ILayerVersion;
      powerToolsLayer: lambda.ILayerVersion;
    };
  };
  leaderboardApi: {
    leaderboardConfigObjectType: ObjectType;
    leaderboardConfigInputype: InputType;
  };
  landingPageApi: {
    landingPageConfigObjectType: ObjectType;
    landingPageConfigInputType: InputType;
  };
  eventbus: EventBus;
}
export class EventsManager extends Construct {
  constructor(scope: Construct, id: string, props: EventsManagerProps) {
    super(scope, id);

    const stack = Stack.of(this);

    const eventsTable = new dynamodb.Table(this, 'EventsTable', {
      partitionKey: {
        name: 'eventId',
        type: dynamodb.AttributeType.STRING,
      },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      encryption: dynamodb.TableEncryption.AWS_MANAGED,
      removalPolicy: RemovalPolicy.DESTROY,
      stream: dynamodb.StreamViewType.NEW_IMAGE,
      pointInTimeRecovery: true,
    });

    const ddbstreamToEventBridgeFunction = new StandardLambdaPythonFunction(this, 'ddbStreamToEvbFunction', {
      entry: 'lib/lambdas/events_ddb_stream_to_evb_function/',
      description: 'Events - DDB stream to EVB',
      index: 'index.py',
      handler: 'lambda_handler',
      timeout: Duration.minutes(1),
      runtime: props.lambdaConfig.runtime,
      memorySize: 128,
      bundling: { image: props.lambdaConfig.bundlingImage },
      layers: [props.lambdaConfig.layersConfig.powerToolsLayer, props.lambdaConfig.layersConfig.helperFunctionsLayer],

      environment: {
        EVENT_BUS_NAME: props.eventbus.eventBusName,
        POWERTOOLS_SERVICE_NAME: 'events_ddb_stream_to_evb',
        LOG_LEVEL: props.lambdaConfig.layersConfig.powerToolsLogLevel,
      },
    });
    props.eventbus.grantPutEventsTo(ddbstreamToEventBridgeFunction.grantPrincipal);
    ddbstreamToEventBridgeFunction.addEventSource(
      new DynamoEventSource(eventsTable, {
        startingPosition: StartingPosition.LATEST,
        batchSize: 1,
      })
    );

    const eventsFunction = new StandardLambdaPythonFunction(this, 'eventsFunction', {
      entry: 'lib/lambdas/events_api/',
      description: 'Events Resolver',
      index: 'index.py',
      handler: 'lambda_handler',
      timeout: Duration.minutes(1),
      runtime: props.lambdaConfig.runtime,
      memorySize: 128,
      bundling: { image: props.lambdaConfig.bundlingImage },
      layers: [props.lambdaConfig.layersConfig.powerToolsLayer, props.lambdaConfig.layersConfig.helperFunctionsLayer],

      environment: {
        DDB_TABLE: eventsTable.tableName,
        POWERTOOLS_SERVICE_NAME: 'events_resolver',
        LOG_LEVEL: props.lambdaConfig.layersConfig.powerToolsLogLevel,
      },
    });

    eventsTable.grantReadWriteData(eventsFunction);

    const eventsDataSourceDdb = props.appsyncApi.api.addDynamoDbDataSource('EventsDataSourceDdb', eventsTable);
    eventsTable.grantReadWriteData(eventsDataSourceDdb);

    // Define the data source for the API
    const eventsDataSource = props.appsyncApi.api.addLambdaDataSource('EventsDataSource', eventsFunction);

    NagSuppressions.addResourceSuppressions(
      eventsDataSource,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Suppress wildcard that covers Lambda aliases in resource path',
          appliesTo: [
            {
              regex: '/^Resource::(.+):\\*$/g',
            },
          ],
        },
      ],
      true
    );

    // Define API Schema
    const trackTypeMethodEnum = new EnumType('TrackType', {
      definition: ['REINVENT_2018', 'REINVENT_2019', 'SUMMIT_SPEEDWAY', 'ATOZ_SPEEDWAY', 'OTHER'],
    });
    props.appsyncApi.schema.addType(trackTypeMethodEnum);

    const raceRankingMethodEnum = new EnumType('RankingMethod', {
      definition: ['BEST_LAP_TIME'],
    });
    props.appsyncApi.schema.addType(raceRankingMethodEnum);

    const raceConfigObjectType = new ObjectType('RaceConfig', {
      definition: {
        raceTimeInMin: GraphqlType.int(),
        numberOfResetsPerLap: GraphqlType.int(),
        trackType: trackTypeMethodEnum.attribute(),
        rankingMethod: raceRankingMethodEnum.attribute(),
        maxRunsPerRacer: GraphqlType.string(),
      },
      directives: [Directive.cognito('admin', 'commentator', 'operator')],
    });
    props.appsyncApi.schema.addType(raceConfigObjectType);

    const raceConfigInputType = new InputType('RaceInputConfig', {
      definition: {
        raceTimeInMin: GraphqlType.int(),
        numberOfResetsPerLap: GraphqlType.int(),
        trackType: trackTypeMethodEnum.attribute(),
        rankingMethod: raceRankingMethodEnum.attribute(),
        maxRunsPerRacer: GraphqlType.string(),
      },
      directives: [Directive.cognito('admin', 'operator')],
    });
    props.appsyncApi.schema.addType(raceConfigInputType);

    const trackObjectType = new ObjectType('Track', {
      definition: {
        trackId: GraphqlType.id(),
        fleetId: GraphqlType.id(),
        leaderBoardTitle: GraphqlType.string(),
        leaderBoardFooter: GraphqlType.string(),
      },
      directives: [Directive.cognito('admin', 'commentator', 'operator')],
    });
    props.appsyncApi.schema.addType(trackObjectType);

    const trackInputType = new InputType('TrackInput', {
      definition: {
        trackId: GraphqlType.id({ isRequired: true }),
        fleetId: GraphqlType.id(),
        leaderBoardTitle: GraphqlType.string({ isRequired: true }),
        leaderBoardFooter: GraphqlType.string({ isRequired: true }),
      },
      directives: [Directive.cognito('admin', 'operator')],
    });
    props.appsyncApi.schema.addType(trackInputType);

    const typeOfEventEnum = new EnumType('TypeOfEvent', {
      definition: ['PRIVATE_WORKSHOP', 'PRIVATE_TRACK_RACE', 'OFFICIAL_WORKSHOP', 'OFFICIAL_TRACK_RACE', 'OTHER'],
    });
    props.appsyncApi.schema.addType(typeOfEventEnum);

    const eventObjectType = new ObjectType('Event', {
      definition: {
        eventId: GraphqlType.id(),
        createdAt: GraphqlType.awsDateTime(),
        createdBy: GraphqlType.id(),
        eventName: GraphqlType.string(),
        typeOfEvent: typeOfEventEnum.attribute(),
        eventDate: GraphqlType.awsDate(),
        sponsor: GraphqlType.string(),
        countryCode: GraphqlType.string(),
        raceConfig: raceConfigObjectType.attribute(),
        tracks: trackObjectType.attribute({ isList: true }),
        landingPageConfig: props.landingPageApi.landingPageConfigObjectType.attribute(),
      },
      directives: [Directive.cognito('admin', 'operator', 'commentator', 'registration')],
    });

    props.appsyncApi.schema.addType(eventObjectType);

    // Event methods
    props.appsyncApi.schema.addQuery(
      'getEvents',
      new ResolvableField({
        returnType: eventObjectType.attribute({ isList: true }),
        dataSource: eventsDataSourceDdb,
        requestMappingTemplate: appsync.MappingTemplate.dynamoDbScanTable(),
        responseMappingTemplate: appsync.MappingTemplate.dynamoDbResultList(),
        directives: [Directive.cognito('admin', 'operator', 'commentator', 'registration')],
      })
    );

    props.appsyncApi.schema.addMutation(
      'addEvent',
      new ResolvableField({
        args: {
          eventName: GraphqlType.string({ isRequired: true }),
          typeOfEvent: typeOfEventEnum.attribute({ isRequired: true }),
          tracks: trackInputType.attribute({ isRequiredList: true }),
          eventDate: GraphqlType.awsDate(),
          sponsor: GraphqlType.string(),
          countryCode: GraphqlType.string(),
          raceConfig: raceConfigInputType.attribute({ isRequired: true }),
          landingPageConfig: props.landingPageApi.landingPageConfigInputType.attribute(),
        },
        returnType: eventObjectType.attribute(),
        dataSource: eventsDataSource,
        directives: [Directive.cognito('admin', 'operator')],
      })
    );

    props.appsyncApi.schema.addSubscription(
      'onAddedEvent',
      new ResolvableField({
        returnType: eventObjectType.attribute(),
        dataSource: props.appsyncApi.noneDataSource,
        requestMappingTemplate: appsync.MappingTemplate.fromString(
          `{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }`
        ),
        responseMappingTemplate: appsync.MappingTemplate.fromString('$util.toJson($context.result)'),
        directives: [Directive.subscribe('addEvent'), Directive.cognito('admin', 'commentator', 'operator')],
      })
    );

    props.appsyncApi.schema.addMutation(
      'deleteEvents',
      new ResolvableField({
        args: { eventIds: GraphqlType.string({ isRequiredList: true }) },
        returnType: GraphqlType.awsJson({ isList: true }),
        dataSource: eventsDataSource,
        directives: [Directive.cognito('admin', 'operator')],
      })
    );
    props.appsyncApi.schema.addSubscription(
      'onDeletedEvents',
      new ResolvableField({
        returnType: GraphqlType.awsJson({ isList: true }),
        dataSource: props.appsyncApi.noneDataSource,
        requestMappingTemplate: appsync.MappingTemplate.fromString(
          `{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }`
        ),
        responseMappingTemplate: appsync.MappingTemplate.fromString('$util.toJson($context.result)'),
        directives: [Directive.subscribe('deleteEvents'), Directive.cognito('admin', 'commentator', 'operator')],
      })
    );

    props.appsyncApi.schema.addMutation(
      'updateEvent',
      new ResolvableField({
        args: {
          eventId: GraphqlType.string({ isRequired: true }),
          eventName: GraphqlType.string({ isRequired: true }),
          typeOfEvent: typeOfEventEnum.attribute({ isRequired: true }),
          tracks: trackInputType.attribute({ isRequiredList: true }),
          eventDate: GraphqlType.awsDate(),
          sponsor: GraphqlType.string(),
          countryCode: GraphqlType.string(),
          raceConfig: raceConfigInputType.attribute({ isRequired: true }),
          landingPageConfig: props.landingPageApi.landingPageConfigInputType.attribute(),
        },
        returnType: eventObjectType.attribute(),
        dataSource: eventsDataSource,
        directives: [Directive.cognito('admin', 'operator')],
      })
    );
    props.appsyncApi.schema.addSubscription(
      'onUpdatedEvent',
      new ResolvableField({
        returnType: eventObjectType.attribute(),
        dataSource: props.appsyncApi.noneDataSource,
        requestMappingTemplate: appsync.MappingTemplate.fromString(
          `{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }`
        ),
        responseMappingTemplate: appsync.MappingTemplate.fromString('$util.toJson($context.result)'),
        directives: [Directive.subscribe('updateEvent'), Directive.cognito('admin', 'commentator', 'operator')],
      })
    );
  }
}
